ray-curator/ray_curator/stages/audio/inference/asr_nemo.py

```diff
@@ -0,0 +1,145 @@
+    cuda: str = ""
+    name: str = "ASR_inference"
+    _start_time = time.time()
+    _metrics: ClassVar[dict] = {}
+    _batch_size: int = 16
+    _resources: Resources = field(default_factory=lambda: Resources(cpus=1.0))
+
+    def check_cuda(self) -> torch.device:
+        self.cuda = str(self.cuda)
+        if self.cuda != "":
+            map_location = torch.device(f"cuda:{self.cuda}") if self.cuda.isdigit() else torch.device(self.cuda)
+        elif torch.cuda.is_available():
+            map_location = torch.device("cuda:0")
+        else:
+            map_location = torch.device("cpu")
+        return map_location
+
+    def setup_on_node(self, _node_info: NodeInfo | None = None, _worker_metadata: WorkerMetadata = None) -> None:
+        # TODO: load asr_model file only once per node
+        pass
+
+    def setup(self, _worker_metadata: WorkerMetadata = None) -> None:
+        """Initialise heavy object self.asr_model: nemo_asr.models.ASRModel"""
+        try:
+            map_location = self.check_cuda()
+            self.asr_model = nemo_asr.models.ASRModel.from_pretrained(
+                model_name=self.model_name, map_location=map_location
+            )
+        except Exception as e:
+            msg = f"Failed to download {self.model_name}"
+            raise RuntimeError(msg) from e
+
+    def inputs(self) -> tuple[list[str], list[str]]:
+        """Define the input attributes required by this stage.
+
+        Returns:
+            Tuple of (top_level_attrs, data_attrs) where:
+            - top_level_attrs: ["data"] - requires FileGroupTask.data to be populated
+        """
+        return ["data"], []
+
+    def outputs(self) -> tuple[list[str], list[str]]:
+        """Define the output attributes produced by this stage.
+
+        Returns:
+            Tuple of (top_level_attrs, data_attrs) where:
+            - top_level_attrs: ["data"] - populates FileGroupTask.data
+            - data_attrs: [self.filepath_key, self.pred_text_key] - audiofile path and predicted text.
+        """
+        return ["data"], [self.filepath_key, self.pred_text_key]
+
+    def transcribe(self, files: list[str]) -> list[str]:
+        """Run inference for speech recognition model
+         Args:
+            files: list of audio file paths.
+
+        Returns:
+            list of predicted texts.
+        """
+        outputs = self.asr_model.transcribe(files)
+        return [output.text for output in outputs]
+
+    def process_batch(self, tasks: list[FileGroupTask | DocumentBatch | SpeechObject]) -> list[SpeechObject]:
+        """Process a audio task by reading audio file and do ASR inference.
+
+
+        Args:
+            tasks: List of FileGroupTask containing a path to audop file for inference.
+
+        Returns:
+            List of SpeechObject with self.filepath_key .
+            If errors occur, the task is returned with error information stored.
+        """
+
+        if isinstance(tasks[0], FileGroupTask):
```

Comment:
In addition to checking the first item of the list, can you add a check that all list items are the same type?

Reply:
done